ami/models/components/sioconv_multi_lr.py

```diff
@@ -1,50 +1,47 @@
+import numpy as np
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
 from torch import Tensor
-import numpy as np
 
 from .sioconvps import RMSNorm, SioConvPSLayer
 from .stacked_hidden_state import StackedHiddenState
 
+
 class MultiLRMLP(nn.Module):
-    def __init__(self, dim: int, dim_ff_hidden: int, num_head: int, lr_scale: float, weight_decay: float, activation: nn.Module=nn.SiLU()):
+    def __init__(
+        self, dim: int, dim_ff_hidden: int, lr_scale: float, weight_decay: float, activation: nn.Module = nn.SiLU()
+    ):
         super().__init__()
-        assert dim_ff_hidden % num_head == 0
-        head_dim = dim_ff_hidden // num_head
         self.fc1_weight = nn.Parameter(torch.randn(dim_ff_hidden, dim))
         self.fc1_bias = nn.Parameter(torch.zeros(dim_ff_hidden))
         self.fc2_weight = nn.Parameter(torch.randn(dim, dim_ff_hidden))
         self.fc2_bias = nn.Parameter(torch.zeros(dim))
         self.activation = activation
 
-        self.lr_scale = torch.exp(torch.linspace(0, np.log(lr_scale), num_head))
-        self.weight_decay = torch.exp(np.log(weight_decay) + torch.linspace(-np.log(lr_scale), 0, num_head))
+        self.lr_scale = torch.exp(torch.linspace(0, np.log(lr_scale), dim_ff_hidden))
+        self.weight_decay = torch.exp(np.log(weight_decay) + torch.linspace(-np.log(lr_scale), 0, dim_ff_hidden))
 
-        self.fc1_weight.register_hook(lambda grad: (
-            grad.view(num_head, head_dim, dim).mul_(self.lr_scale[:,None,None]) +
-            self.fc1_weight.view(num_head, head_dim, dim).mul_(self.weight_decay[:,None,None]))
-        .view(dim_ff_hidden, dim))
-        self.fc1_bias.register_hook(lambda grad: (
-            grad.view(num_head, head_dim).mul_(self.lr_scale[:,None]) +
-            self.fc1_bias.view(num_head, head_dim).mul_(self.weight_decay[:,None]))
-        .view(dim_ff_hidden))
-        self.fc2_weight.register_hook(lambda grad: (
-            grad.view(dim, num_head, head_dim).mul_(self.lr_scale[None,:,None]) +
-            self.fc2_weight.view(dim, num_head, head_dim).mul_(self.weight_decay[None,:,None]))
-        .view(dim, dim_ff_hidden))
+        self.fc1_weight.register_hook(
+            lambda grad: grad.mul_(self.lr_scale[:, None]) + self.fc1_weight.mul_(self.weight_decay[:, None])
+        )
+        self.fc1_bias.register_hook(lambda grad: grad.mul_(self.lr_scale) + self.fc1_bias.mul_(self.weight_decay))
+        self.fc2_weight.register_hook(
+            lambda grad: grad.mul_(self.lr_scale[None, :]) + self.fc2_weight.mul_(self.weight_decay[None, :])
+        )
```

Comment:
`hook`メソッドの引数テンソル自体を変更してしまう `mul_` (アンダースコア付き演算)は非推奨です。
```suggestion
        self.fc1_weight.register_hook(
            lambda grad: grad.mul(self.lr_scale[:, None]) + self.fc1_weight.mul(self.weight_decay[:, None])
        )
        self.fc1_bias.register_hook(lambda grad: grad.mul(self.lr_scale) + self.fc1_bias.mul(self.weight_decay))
        self.fc2_weight.register_hook(
            lambda grad: grad.mul(self.lr_scale[None, :]) + self.fc2_weight.mul(self.weight_decay[None, :])
        )
```

> The hook should not modify its argument, but it can optionally return a new gradient which will be used in place of [grad](https://pytorch.org/docs/stable/generated/torch.Tensor.grad.html#torch.Tensor.grad).

https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html

Reply:
https://github.com/MLShukai/ami/pull/292/commits/7706641103427c0a13a6b172292dcb7b3bb24045
修正しました